privacyidea/lib/policydecorators.py

```diff
@@ -156,28 +156,41 @@ def auth_cache(wrapped_function, user_object, passw, options=None):
     options = options or {}
     g = options.get("g")
     auth_cache_dict = None
+
     if g:
         auth_cache_dict = Match.user(g, scope=SCOPE.AUTH, action=ACTION.AUTH_CACHE,
                                      user_object=user_object).action_values(unique=True, write_to_audit_log=False)
         if auth_cache_dict:
+
             # verify in cache and return an early success
             auth_times = list(auth_cache_dict)[0].split("/")
             # determine first_auth from policy!
             first_offset = parse_timedelta(auth_times[0])
 
-            if len(auth_times) == 2:
-                # Determine last_auth from policy
-                last_offset = parse_timedelta(auth_times[1])
+            import re
+            # Use auth cache when number of allowed authentications is defined
+            if len(auth_times) == 2 and re.match(r"^\d+$", auth_times[1]):
```

Comment:
Since we have the same test below, this block should be moved there. Makes the flow a little clearer.

Reply:
Honestly, I am not sure how that could be done. This is the only thing I could come up with:
```
last_offset = None
max_number_of_authentications = None
if len(auth_times) == 2:
    if re.match(r"^\d+$", auth_times[1]):
        # Check if second entry defines max number of authentications
        max_number_of_authentications = int(auth_times[1])
    else:
        last_offset = parse_timedelta(auth_times[1])
else:
    # No second value is defined in auth_times
    last_offset = first_offset

last_auth = None
if last_offset:
    last_auth = datetime.datetime.utcnow() - last_offset

result = verify_in_cache(user_object.login, user_object.realm,
                         user_object.resolver, passw,
                         first_auth=first_auth,
                         last_auth=last_auth,
                         max_number_of_authentications=max_number_of_authentications)
```
and it is not good to read.

Reply:
You could get rid of the last `if:` when You calculate `last_auth` directly:
```
            # Use auth cache when number of allowed authentications is defined
            if len(auth_times) == 2:
                if re.match(r"^\d+$", auth_times[1]):
                    max_auths = int(auth_times[1])
                else:
                    # Determine last_auth delta from policy
                    last_offset = parse_timedelta(auth_times[1])
                    last_auth = datetime.datetime.utcnow() - last_offset
            else:
                # If there is no last_auth, it is equal to first_auth
                last_auth = datetime.datetime.utcnow() - first_offset

            result = verify_in_cache(...)
```